bot/exts/valentines/markov_poem_generator.py

```diff
@@ -0,0 +1,513 @@
+
+import async_timeout
+import markovify
+from discord import Embed
+from discord.ext import commands
+
+from bot.constants import Colours
+from bot.utils.pagination import LinePaginator
+
+log = logging.getLogger(__name__)
+
+_EXECUTOR = ThreadPoolExecutor(10)
+
+
+async def in_thread(func: Callable) -> Future:
+    """Allows non-async functions to work in async functions."""
+    loop = get_event_loop()
+    return await loop.run_in_executor(_EXECUTOR, func)
+
+
+class MakeShortSentenceRanOut(commands.CommandError):
+    """Raised when the markov chain can no longer output sentences."""
+
+    pass
+
+
+class RhymingSentenceNotFound(commands.CommandError):
+    """Raised when a rhyming sentence could not be found."""
+
+    def __init__(
+        self,
+        scheme: str,
+        unit_count: Dict[str, int],
+        time_start: datetime.time
+    ):
+        self.scheme = scheme
+        self.unit_count = unit_count
+        self.time_start = time_start
+
+
+class RhymeAPIUnresponsive(commands.CommandError):
+    """Raised when the API does not return 200."""
+
+    pass
+
+
+class Cache:
+    """A context manager to facilitate the storage of word to rhyme sets."""
+
+    cache: Dict[str, Set[str]] = {}
+
+    def __init__(self, word: str):
+        self.word = word
+
+    def __enter__(self) -> Tuple[Dict[str, Set[str]], str]:
+        self.word = self.word.replace("'", "e")  # Sometimes ' replaces e
+        return self.cache, self.word
+
+    def __exit__(
+        self,
+        exc_type: Optional[Type[BaseException]],
+        exc_value: Optional[BaseException],
+        traceback: Optional[TracebackType]
+    ):
+        # If an error is raised, then the word may not be added to the cache.
+        if self.word in self.cache and len(self.cache[self.word]) == 0:
+            logging.info(f"No rhymes were found for the word: {self.word}")
+
+
+def memoize(func: Callable) -> Callable:
+    """
+    A decorator to access and cache rhyme sets.
+
+    If the rhyme set of the word already exists in the cache, then
+    there is no need to execute `func` because results can be taken from the
+    cache. Otherwise, execute `func` and store the results into the cache.
+    """
+    @functools.wraps(func)
+    async def wrapper(
+        *args,
+        instance: Callable = None,
+        word: str = None
+    ) -> Set[str]:
+        with Cache(word) as (cache, word):
+            if word not in cache.keys():
+                cache[word] = await func(instance, word)
+
+        return cache[word]
+
+    return wrapper
+
+
+def command_wrapper(func: Callable) -> Callable:
+    """The decorator adds a timeout and shows the bot's `typing...` text."""
+    @functools.wraps(func)
+    async def wrapper(
+        *args,
+        instance: Callable = None,
+        timeout: int = None,
+        ctx: commands.Context = None,
+        scheme: str = None,
+        unit_count: Dict[str, int] = None,
+        time_start: datetime.time = None,
```

Comment:
The type annotations with `datetime.time` are incorrect, as you have done a `from datetime import datetime`; so the [`datetime.time` here refers to the _method_ `time`](https://docs.python.org/3/library/datetime.html#datetime.datetime.time) of `datetime.datetime` objects, and not the [`datetime.time`](https://docs.python.org/3/library/datetime.html#datetime.time) class